query-algebrizer/src/clauses/not.rs

```diff
@@ -0,0 +1,450 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use std::collections::btree_map::Entry;
+
+use mentat_core::{
+    Schema,
+};
+
+use mentat_query::{
+    NotJoin,
+    UnifyVars,
+    ContainsVariables,
+};
+
+use clauses::ConjoiningClauses;
+
+use errors::{
+    Result,
+};
+
+use types::{
+    ColumnConstraint,
+    ComputedTable,
+    EmptyBecause,
+};
+
+impl ConjoiningClauses {
+    pub fn apply_not_join(&mut self, schema: &Schema, not_join: NotJoin) -> Result<()> {
+        let projected = match not_join.unify_vars {
+            UnifyVars::Implicit => not_join.collect_mentioned_variables(),
+            UnifyVars::Explicit(vs) => vs.into_iter().collect(),
+        };
+        let mut template = self.use_as_template(&projected);
+        let mut empty_because: Option<EmptyBecause> = None;
+        
+        for clause in not_join.clauses.into_iter() {
+            template.apply_clause(&schema, clause)?;
+            // This is disgusting, but I can't think of anything else right now that works.
+            if template.is_known_empty() {
+                empty_because = template.empty_because;
+                template.empty_because = None;
+            }
+        }
+
+        if template.wheres.is_empty() && empty_because.is_some() {
+            return Ok(());
+        } else {
+            for (v, cols) in (&self.column_bindings).iter().filter(|&(key, _)| projected.contains(&key) ) {
+                match template.column_bindings.entry(v.clone()) {
+                    Entry::Vacant(e) => {
+                        e.insert(cols.clone());
+                    },
+                    Entry::Occupied(mut e) => {
+                        e.get_mut().append(&mut cols.clone());
+                    },
+                }
+            }
+        }
+
+        // We are only expanding column bindings here and not pruning extracted types as we are not projecting values.
```

Comment:
So, this is what I am getting out for this query from the code that I currently have:
```
SELECT 
    DISTINCT `datoms00`.e AS `?x` 
FROM 
    `datoms` AS `datoms00`, 
    `all_datoms` AS `all_datoms01` 
WHERE 
    `datoms00`.a = 99 
    AND NOT EXISTS (
        SELECT 1 
        FROM `all_datoms` AS `all_datoms02` 
        WHERE `all_datoms02`.v = `all_datoms01`.v LIMIT 1) 
    AND `datoms00`.e = `all_datoms01`.e
```

It seems that the differences here are referring to `alldatoms` instead of `datoms` for `datoms01` & `datoms02`, and linking in the subquery through `datomsX.v` instead of `datomsX.value_type_tag`. Can I confirm that they are the right places to be concentrating?